File: medium/3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.rb
# @param {String} s
# @return {Integer}

def length_of_longest_substring(s)
    set = [].to_set
    l = 0
    r = 0
    result = 0
    
    while r < s.length do
        if set.include? s[r]
            while set.include? s[r] do 
                set.delete s[l]
                l += 1
            end
        end
        set.add s[r]
        r += 1
        result = [result, set.size].max
    end
    
    result
end

File: medium/3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
const lengthOfLongestSubstring = function(s) {
    let set = new Set(),
        res = 0,
        l = 0;
    for (let r = 0; r < s.length; r++) {
        if (set.has(s[r])) {
            while (set.has(s[r])) {
                set.delete(s[l]);
                l++;
            };
        }
        set.add(s[r]);
        res = Math.max(res, set.size);
    }
    return res;
};